### DISCUSSION TITLE
Passing Bearer token in header

### DISCUSSION THREAD

**Discussion by ndenic** [pactumjs/pactum#179] [Q&A]
Hello folks,

i recently started using PactumJS and now i need to pass token in header which i'm getting from first post request

```js
  it.only('Sync settings and validate JSON scehma', async () => {
        const tokence = await pactum.spec()
        .withHeaders('Content-type', 'application/json')
        .withBody({"username": "user1", "password": "user1"})
        .post(endpoints.auth)
        .expectStatus(200)
        .returns('meta.access_token')


        await pactum.spec()
        .withHeaders({
            'Authorization': `Bearer ${tokence}`,
            'Content-Type': 'application/json'
        })
        .post(endpoints.sync) 
        .expectStatus(200)
    })
    
```
    
Everything is okay in first part of grabbing access token, but when i pass that token in second part spec()... im getting that "JWT token expired or invalid" .What could be the problem, I would appreciate your help. Thanks in advance
Part of header looks like this
```json
 "headers": {
    "Authorization": "Bearer GOOD TOKEN (deleted from security reasons)",
    "Content-Type": "application/json",
    "Accept": "*/*"
  },
```
```json
 {
        "err_key": "unauthorized",
        "message": "JWT token expired or invalid",
        "details": ""
    }
```
error message is 401

**Comment by ASaiAnudeep** [comment]
I don't find any issue with your test case. Are you able to use the same token in Postman?

**Reply by ndenic** [reply]
Yes , i tried with Postman and it works - with same token

**Reply by ASaiAnudeep** [reply]
Try using all the headers that are part of Postman request in your pactum test.

**Reply by ndenic** [reply]
It didn't help and i added all that i could, but thanks anyway. 
Also tried to decode that token with https://jwt.io/ and everything is okay with token.

**Reply by ASaiAnudeep** [reply]
Interesting. Ideally if all the headers are right it should work. Try `inspecting` all the request params. I hope you don't have any default headers.

```js
await pactum.spec()
  .withHeaders({
    'Authorization': `Bearer ${tokence}`,
    'Content-Type': 'application/json'
  })
  .post(endpoints.sync)
  .inspect()
```

**Reply by ndenic** [reply]
This is the result. Weird thing that first part with login pass and its grabbing right token

[W] Request
{
  "headers": {
    "Authorization": "Bearer GOOD TOKEN",
    "Accept": " */* ", - its * / *
    "Connection": "keep-alive",
    "Accept-Encoding": "gzip, deflate, br",
    "Host": "i delete this one"
  },
  "url": "good url",
  "method": "POST",
  "path": "good path",
  "timeout": 3000
}
[W] Response
{
  "statusCode": 401,
  "headers": {
    "date": "Fri, 08 Jul 2022 12:51:27 GMT",
    "server": "Apache/2.4.38 (Debian)",
    "access-control-allow-origin": "*",
    "access-control-allow-methods": "GET,PUT,POST,DELETE,OPTIONS",
    "access-control-allow-headers": "Content-Type, Origin, x-origin, authorization, Cache-Control, *",
    "access-control-allow-credentials": "true",
    "x-php-version": "7.4.29",
    "x-phalcon-version": "4.1.0",
    "status": "401 Unauthorized",
    "content-length": "120",
    "content-type": "application/json; charset=UTF-8",
    "keep-alive": "timeout=5, max=100",
    "connection": "Keep-Alive"
  },
  "body": {
    "errors": [
      {
        "err_key": "unauthorized",
        "message": "JWT token expired or invalid",
        "details": ""
      }
    ],
    "meta": {
      "status_code": 401
    }
  }
}

**Reply by samuelvinay91** [reply]
Hi @ASaiAnudeep - Any update on this even I was trying the same to authenticate using username/password and sign with with the token. i am getting bad csrf.But it seems like it is correct

**Reply by samuelvinay91** [reply]
I was able to authenticate further requests using supertest.

**Reply by ASaiAnudeep** [reply]
I have been using bearer token in multiple projects which is working perfectly fine.

If you can provide steps to reproduce the issue, it would be helpful to investigate.

**Reply by samuelvinay91** [reply]
@ASaiAnudeep - Thank you so much for your reply.

This below code using axios was working fine for all subsequent calls internally I am using agent

```
  const cookiejar = new tough.CookieJar();
    agent = await axios.create({
    jar: cookiejar,
    withCredentials: true,
   });

```

```
let res = await agent.get("/signin");
    agent.defaults.headers["x-csrf-token"] = res.headers["x-csrf-token"];
    axios.defaults.headers.post["X-CSRF-Token"] = res.headers["x-csrf-token"];
    axios.defaults.headers.post["Content-Type"] =
      "application/x-www-form-urlencoded";
    let k = await agent.post("/signin", {
      email: process.env.MFA_USERNAME,
      password: process.env.MFA_PASSWORD,
      _csrf: res.headers["x-csrf-token"],
    });
```

same I am trying to do in pactum

```
  let res = await pactum.spec().get("/signin");
    console.log(res.headers);
    cookie = res.headers["set-cookie"][0];
    // agent.defaults.headers["x-csrf-token"] = res.headers["x-csrf-token"];
    // axios.defaults.headers.post["X-CSRF-Token"] = res.headers["x-csrf-token"];
    // axios.defaults.headers.post["Content-Type"] =
    //   "application/x-www-form-urlencoded";
    let k = await pactum.spec().post("/signin").withJson({
      email: process.env.MFA_USERNAME,
      password: process.env.MFA_PASSWORD,
      _csrf: res.headers["x-csrf-token"],
    }).expectStatus(200).inspect();
    cookie = k.headers["set-cookie"][0];
    console.log(" >> ", k.status, cookie);
```
Correct me please if I am missing or doing anything wrong

**Reply by ASaiAnudeep** [reply]
Is there an Open API where we can test this code?

**Reply by samuelvinay91** [reply]
@ASaiAnudeep - i doubt we don't have

**Reply by samuelvinay91** [reply]
@ASaiAnudeep - I just needed a way to getCookiesandCsrf token authenticated and continue with following API calls using in same session .kindly suggest

**Reply by ASaiAnudeep** [reply]
If I could replicate the issue, it would be easy for me to fix it.

Can you try making the same request using https://www.npmjs.com/package/phin
and https://www.npmjs.com/package/agent-phin

**Reply by ASaiAnudeep** [reply]
@SamuelVinay91 Can you create a new discussion to follow up on this. This discussion is related to jwt token.
